simpeg_drivers/utils/regularization.py

```diff
@@ -0,0 +1,362 @@
+# '''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+#  Copyright (c) 2025 Mira Geoscience Ltd.                                          '
+#                                                                                   '
+#  This file is part of simpeg-drivers package.                                     '
+#                                                                                   '
+#  simpeg-drivers is distributed under the terms and conditions of the MIT License  '
+#  (see LICENSE file at the root of this source code package).                      '
+#                                                                                   '
+# '''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+import numpy as np
+import scipy.sparse as ssp
+from discretize import TreeMesh
+from simpeg.utils import mkvc, sdiag
+
+
+def cell_neighbors_along_axis(mesh: TreeMesh, axis: str) -> np.ndarray:
+    """
+    Get adjacent cells along provided axis for all cells in the mesh
+
+    :param mesh: Input TreeMesh.
+    :param axis: Cartesian axis along which to find neighbors.  Must be
+        'x', 'y', or 'z'.
+    """
+
+    if axis not in "xyz":
+        raise ValueError("Argument 'axis' must be one of 'x', 'y', or 'z'.")
+
+    stencil = getattr(mesh, f"cell_gradient_{axis}")
+    ith_neighbor, jth_neighbor, _ = ssp.find(stencil)
+    n_stencils = int(ith_neighbor.shape[0] / 2)
+    stencil_indices = jth_neighbor[np.argsort(ith_neighbor)].reshape((n_stencils, 2))
+
+    return np.sort(stencil_indices, axis=1)
+
+
+def cell_neighbors(mesh: TreeMesh) -> np.ndarray:
+    """Find all cell neighbors in a TreeMesh."""
+
+    x_neighbors = cell_neighbors_along_axis(mesh, "x")
+    x_neighbors_backward = np.fliplr(x_neighbors)
+    y_neighbors = cell_neighbors_along_axis(mesh, "y")
+    y_neighbors_backward = np.fliplr(y_neighbors)
+    max_index = np.max([x_neighbors.max(), y_neighbors.max()])
+    if mesh.dim == 3:
+        z_neighbors = cell_neighbors_along_axis(mesh, "z")
+        z_neighbors_backward = np.fliplr(z_neighbors)
+        max_index = np.max([max_index, z_neighbors.max()])
+
+    all_neighbors = []  # Store
+    x_adjacent = np.ones(max_index + 1, dtype="int") * -1
+    y_adjacent = np.ones(max_index + 1, dtype="int") * -1
+    x_adjacent_backward = np.ones(max_index + 1, dtype="int") * -1
+    y_adjacent_backward = np.ones(max_index + 1, dtype="int") * -1
+
+    x_adjacent[y_neighbors[:, 0]] = y_neighbors[:, 1]
+    y_adjacent[x_neighbors[:, 1]] = x_neighbors[:, 0]
+
+    x_adjacent_backward[y_neighbors_backward[:, 0]] = y_neighbors_backward[:, 1]
+    y_adjacent_backward[x_neighbors_backward[:, 1]] = x_neighbors_backward[:, 0]
```

Comment:
```suggestion
    x_adjacent[y_neighbors[:, 0]] = y_neighbors[:, 1]
    y_adjacent[x_neighbors[:, 1]] = x_neighbors[:, 0]

    x_adjacent_backward[y_neighbors_backward[:, 0]] = y_neighbors_backward[:, 1]
    y_adjacent_backward[x_neighbors_backward[:, 1]] = x_neighbors_backward[:, 0]
```
There is duplication here. We could make a sub-function that only deals with getting the adjacent corners, and call it twice with just a `np.fliplr` on the neighbour arrays.

Reply:
I started a refactor when I began this issue, but backed off fearing I would miss something.  I can take another look at it now that I think i understand the code better!

Reply:
Yea fair enough. Let's get the run test all polished up - we can clean up with confidence after.

Reply:
refactored